extensions/ql-vscode/src/view/common/errors.ts

```diff
@@ -0,0 +1,54 @@
+import { getErrorMessage, getErrorStack } from "../../pure/helpers-pure";
+import { vscode } from "../vscode-api";
+
+// Keep track of previous errors that have happened.
+// The listeners for uncaught errors and rejections can get triggered
+// twice for each error. This is believed to be an effect caused
+// by React's error boundaries. Adding an error boundary stops
+// this duplicate reporting for errors that happen during component
+// rendering, but unfortunately errors from event handlers and
+// timeouts are still duplicated and there does not appear to be
+// a way around this.
+const previousErrors: Set<Error> = new Set();
+
+function shouldReportError(error: Error): boolean {
+  const seenBefore = previousErrors.has(error);
+  previousErrors.add(error);
+  setTimeout(() => {
+    previousErrors.delete(error);
+  }, 1000);
+  return !seenBefore;
+}
+
+const unhandledErrorListener = (event: ErrorEvent) => {
+  if (shouldReportError(event.error)) {
+    vscode.postMessage({
+      t: "unhandledError",
+      error: {
+        message: getErrorMessage(event.error),
+        stack: getErrorStack(event.error),
+      },
+    });
+  }
+};
+
+const unhandledRejectionListener = (event: PromiseRejectionEvent) => {
```

Comment:
Should we be making a distinction between the unhandled errors and unhandled rejections? As far as I am aware, unhandled rejections don't cause any user-visible behaviour (except perhaps some data not loading), but unhandled errors might cause the complete view to disappear. It would be useful to make that distinction in the telemetry.

Reply:
For now I'm going to leave them combined. I don't think it's quite as well defined as saying unhandled errors are always more important. Missing or incorrect data is also bad, and maybe worse if the user don't know the data they're looking at is incorrect. Unhandled errors can also happen outside of rendering.

Also, we can certainly split these up in the future, and doing it after this PR will be fine. The telemetry is quite schema-less and we only keep it for a short period, so changing the format is quite cheap to do.